README.rst
.. image:: https://mybinder.org/badge_logo.svg
 :target: https://mybinder.org/v2/gh/SofianeB/ECAS-B2SHARE/master?urlpath=lab
 
.. image:: https://travis-ci.org/SofianeB/ECAS-B2SHARE.svg?branch=master
   :target: https://travis-ci.org/SofianeB/ECAS-B2SHARE 
    
.. image:: https://readthedocs.org/projects/ecas-b2share/badge/?version=latest
   :target: https://ecas-b2share.readthedocs.io/en/latest/?badge=latest
   :alt: Documentation Status
   
.. image:: https://badge.fury.io/py/ecasb2share.svg
    :target: https://badge.fury.io/py/ecasb2share
   
.. image:: https://coveralls.io/repos/github/SofianeB/ECAS-B2SHARE/badge.svg?branch=master
    :target: https://coveralls.io/github/SofianeB/ECAS-B2SHARE?branch=master

    
============
ECAS B2SHARE
============


Python module to interact with the B2SHARE HTTP REST API!

.. note::

   This module does not cover the whole B2SHARE API and is specific to the `ECAS use case <https://ee-docs.readthedocs.io/en/latest/>`_.


How to install
==============

Using pip
::

   pip install ecasb2share


How to use
==========

Create a client
::

   from ecasb2share.ecasb2shareclient import EcasShare as Client
   client = Client(url, token_file)


Create a draft record
::

   client.create_draft_record_with_pid(title, pid)

Documentation
=============

For more details, please check the technical documentation `here <https://ecas-b2share.readthedocs.io/en/latest/>`_. 


CHANGELOG.rst
=========
Changelog
=========


Version 0.0.1b6 2019-02-19
==========================
- Fix default token path

Version 0.0.1b5 2019-02-14
==========================

- Update README
- Fix uploading token
- Fix creating new drafts records (filebucketid)

Version 0.0.1b4 2019-02-12
==========================

- Fixed bug in adding files to records

Version 0.0.1-a0
================

- Added first basic methods



ecasb2share/tests/__init__.py
all = ["ecas_b2share_test", "tests.py"]

ecasb2share/tests/ecas_b2share_test.py
import unittest
import requests
import json

from ecasb2share.ecasb2shareclient import EcasShare
from ecasb2share.exceptions import MetadataKeyMissingException, MetadataException, PidSyntaxException
from unittest.mock import Mock, patch, mock_open
from nose.tools import assert_is_not_none

RECORD = json.load(open('test_files/record.json'))
RECORD_ID = 'b4da58206da24b1aacf3b35c66024ea8'
RECORD_ID_NO_PID = '359a7e017d164f6383e58bba925c4276'


class EcasShareTestCase(unittest.TestCase):

    def setUp(self):

        self.ecasb2share = EcasShare(token_file='test_files/token.txt')

    @patch("builtins.open", new_callable=mock_open, read_data=json.dumps({"titles": [{"title": "test 1"}],
                                                                          "community": "hkjhk4987dakh"}))
    def load_json_unit_test(self, mock_open):
        """
        Test if json files loaded properly.

        """

        expected_output = {"titles": [{"title": "test 1"}],
                           "community": "hkjhk4987dakh"}

        filename = 'metadata.json'
        actual_output = self.ecasb2share.load_metadata_from_json(filename)

        # Assert filename is file that is opened
        # mock_open.assert_called_with(filename)

        self.assertEqual(expected_output, actual_output)

    @patch('ecasb2share.ecasb2shareclient.requests.get')
    def send_get_request_unit_test(self, mock_get):

        mock_get.return_value.ok = True

        # Send a request to the API server and store the response.
        response = requests.get('https://b2share.eudat.eu/')

        assert_is_not_none(response)

    @patch('ecasb2share.ecasb2shareclient.requests.put')
    def send_put_request_unit_test(self, mock_get):

        mock_get.return_value.ok = True

        # Send a request to the API server and store the response.
        response = requests.get('https://b2share.eudat.eu/')

        assert_is_not_none(response)

    @patch('ecasb2share.ecasb2shareclient.requests.post')
    def send_post_request_unit_test(self, mock_get):

        mock_get.return_value.ok = True

        # Send a request to the API server and store the response.
        response = requests.post('https://b2share.eudat.eu/')

        assert_is_not_none(response)

    @patch('ecasb2share.ecasb2shareclient.requests.get')
    def list_communities_unit_test(self, mock_get):
        expected_response = {
            "hits": {
                "hits": [
                    {
                        "created": "Tue, 18 Oct 2016 08:30:47 GMT",
                        "description": "The big Eudat community. Use this community if no other is suited for you",
                        "id": "e9b9792e-79fb-4b07-b6b4-b9c2bd06d095",
                        "links": {
                            "self": "https://trng-b2share.eudat.eu/api/communities/e9b9792e-79fb-4b07-b6b4-b9c2bd06d095"},
                        "logo": "/img/communities/eudat.png",
                        "name": "EUDAT",
                        "updated": "Tue, 18 Oct 2016 08:30:47 GMT"},
                    ...],
                "total": 11},
            "links": {
                "self": "https://trng-b2share.eudat.eu/api/communities/"}}

        mock_get.return_value = Mock(ok=True)
        mock_get.return_value.json.return_value = expected_response

        response = self.ecasb2share.list_communities()

        # assert_list_equal(response, expected_response)

    @patch('ecasb2share.ecasb2shareclient.requests.get')
    def list_communities_when_none_unit_test(self, mock_get):
        expected_response = {
            "hits": {
                "hits": [
                    {
                        "created": "Tue, 18 Oct 2016 08:30:47 GMT",
                        "description": "The big Eudat community. Use this community if no other is suited for you",
                        "id": "e9b9792e-79fb-4b07-b6b4-b9c2bd06d095",
                        "links": {
                            "self": "https://trng-b2share.eudat.eu/api/communities/e9b9792e-79fb-4b07-b6b4-b9c2bd06d095"},
                        "logo": "/img/communities/eudat.png",
                        "name": "EUDAT",
                        "updated": "Tue, 18 Oct 2016 08:30:47 GMT"},
                    ...],
                "total": 11},
            "links": {
                "self": "https://trng-b2share.eudat.eu/api/communities/"}}

        mock_get.return_value = Mock(ok=True)
        mock_get.return_value.json.return_value = expected_response

        response = self.ecasb2share.list_communities()

        assert_is_not_none(response)

    # This method will be used by the mock to replace requests.get
    def mocked_requests_get(self, *args, **kwargs):
        class MockResponse:
            def __init__(self, json_data, status_code):
                self.json_data = json_data
                self.status_code = status_code

            def json(self):
                return self.json_data

        if args[0] == 'http://someurl.com/test.json':
            return MockResponse({"key1": "value1"}, 200)
        elif args[0] == 'http://someotherurl.com/anothertest.json':
            return MockResponse({"key2": "value2"}, 200)

        return MockResponse(None, 404)

    def server_running_unit_test(self):

        response = requests.get('https://b2share.eudat.eu')
        self.assertEqual(response.status_code, 200)

    def get_list_communities_unit_test(self):
        with patch('ecasb2share.ecasb2shareclient.requests.get') as mock_get:
            mock_get.return_value.ok = True

            communities = self.ecasb2share.list_communities()

            # self.assertEqual(communities, 'Success')

    def get_record_pid_unit_test(self):
        """
        Check if ePIC_PID found and retrieved from record.

        """

        with patch('ecasb2share.ecasb2shareclient.EcasShare.get_specific_record') as mock_request:

            mock_request.return_value = RECORD

            pid = self.ecasb2share.get_record_pid(RECORD_ID)

            self.assertEqual(pid, RECORD['metadata']['ePIC_PID'])

    def get_record_pid_not_existent_unit_test(self):
        """
        Test if exception is raised when ePIC_PID missing
        """

        with self.assertRaises(MetadataKeyMissingException):
            self.ecasb2share.get_record_pid(RECORD_ID_NO_PID)

    def validate_metadata_true_unit_test(self):
        """
        Check if metadata opened and load properly.
        """

        expected_metadata = json.load(open('test_files/fake_metadata.json', 'r'))
        metadata = self.ecasb2share.validate_metadata(metadata_file='test_files/fake_metadata.json')

        self.assertEqual(metadata, expected_metadata)

    def validate_metadata_false_unit_test(self):
        """
        Check if exception is raised when key missing.
        """

        with self.assertRaises(MetadataException):
            self.ecasb2share.validate_metadata(metadata_file='test_files/metadata_missing_key.json')

    def check_pid_syntax_unit_test(self):
        pid = '00.00000/xxxx-yyyy-zzzz'
        response = self.ecasb2share.check_pid_syntax(pid)
        self.assertTrue(response)

    def check_pid_no_slash_unit_test(self):
        """
        Check if exception raises when no slash in PID.

        """

        pid = '00.00000xxxx-yyyy-zzzz'
        with self.assertRaises(PidSyntaxException):
            self.ecasb2share.check_pid_syntax(pid)

    def check_pid_syntax_no_prefix_unit_test(self):
        """
        Check if exception raises when missing a prefix in the PID.

        """

        pid = '/xxxx-yyyy-zzzz'
        with self.assertRaises(PidSyntaxException):
            self.ecasb2share.check_pid_syntax(pid)

    def check_pid_syntax_no_suffix_unit_test(self):
        """
        Check if exception raises when missing a suffix in the PID.

        """

        pid = '00.00000/'
        with self.assertRaises(PidSyntaxException):
            self.ecasb2share.check_pid_syntax(pid)

    def get_filebucketid_from_record_unit_test(self):
        """
        Test if filebukcetid is correctly retrieved from the record.

        """

        # random id
        record_id = 'sadadasdadddd5dd9f1fa4441a79dfa5f270469f8d'

        # patch get_specific_record
        with patch('ecasb2share.ecasb2shareclient.EcasShare.get_specific_record') as mock_process:

            mock_process.return_value = RECORD
            filebucketid = self.ecasb2share.get_filebucketid_from_record(record_id)

            self.assertEqual(filebucketid, 'da7ddd6c-5d14-4986-91aa-d9a46b4138d8')

    def check_total_community_specific_records_unit_test(self):
        """
        Test if specific records are retrieved from the request response.
        """

        with patch('requests.get') as mock_request:

            mock_request.return_value.ok = True
            # mock_request.return_value.text = json.load(open('test_files/communiy_records.json'))

            records = self.ecasb2share.retrieve_community_specific_records(community_id='d2c6e694-0c0a-4884-ad15-ddf498008320')

            self.assertEqual(records['hits']['total'], 3)

    def check_total_community_specific_records_patched_unit_test(self):
        """
        Test if specific records are retrieved from the request response (patched).
        """

        with patch('ecasb2share.ecasb2shareclient.EcasShare.retrieve_community_specific_records') as mock_request:


             mock_request.return_value = json.load(open('test_files/community_records.json'))

             records = self.ecasb2share.retrieve_community_specific_records()
             print('Info is {}'.format(records))
             self.assertEqual(records['hits']['total'], 3)


ecasb2share/exceptions.py
class PidSyntaxException(Exception):
    """
    Raises when Handle syntax not correct.
    """

    def __init__(self, **args):

        self.msg = 'PID syntax not correct'
        self.correct_syntax = args['correct_syntax']
        self.concrete_msg = args['msg']
        self.pid = args['pid']

        if self.pid is not None:
            self.msg = self.msg.replace('andle', 'andle'+self.pid)

        if self.concrete_msg is not None:
            self.msg += ': '+self.concrete_msg

        if self.correct_syntax is not None:
           self.msg += '\n\tExpected: ' + self.correct_syntax

        self.msg += '.'

        super(self.__class__, self).__init__(self.msg)

class MetadataException(Exception):
    """
    Raises when metadata json file not correct.
    """

    def __init__(self, **args):

        self.msg = "Mandatory metadata item missing"

        self.concrete_msg = args['msg']

        if self.concrete_msg is not None:
            self.msg += ':'+self.concrete_msg
        self.msg += '.'

        super(self.__class__, self).__init__(self.msg)

class MetadataKeyMissingException(Exception):
    """
    Raises when key not created for record
    """

    def __init__(self, **args):

        self.msg = "Metadata key not created yet"

        self.concrete_msg = args['msg']

        if self.concrete_msg is not None:
            self.msg += ':'+self.concrete_msg
        self.msg += '.'

        super(self.__class__, self).__init__(self.msg)

ecasb2share/ecasb2shareclient.py
""" Helper class to interact with B2SHARE REST API to create draft records
from ECAS user workspace.
Please refer to https://eudat.eu/services/userdoc/b2share-usage
 for more information.

.. Note:: this module is not considered as B2SHARE client and it is specific
 to ECAS use case.

Author: Sofiane Bendoukha (DKRZ), 2019

"""

import requests
import json
import os
import logging

from requests import Request, Session
from . import exceptions


from urllib.parse import urljoin
from requests.exceptions import HTTPError

logging.basicConfig(level=logging.INFO)


class EcasShare (object):

    """ ECAS B2SHARE main class """

    # Initialize

    def __init__(self, url=None, token_file=None):
        """
        Initialize the client.

        :param url: URL of the B2SHARE instance.  One of the following
        hostnames can be used to identify the B2SHARE instance:

                     * b2share.eudat.eu - the hostname of the production site.
                     * trng-b2share.eudat.eu - the base url of the training
                     site. Use this URL for testing.

        :param token_file: B2SHARE API ACCESS token
        """

        # Default path in container
        CONTAINER_PATH = '/home/jovyan/work/conf'

        # TODO: set up b2share url (prod or test), token and payload?
        if url is None:
            self.B2SHARE_URL = 'https://eudat-b2share-test.csc.fi'
        else:
            self.B2SHARE_URL = url

        if token_file is None:
            self.token_path = os.path.join(CONTAINER_PATH, 'token.txt')
        else:
            self.token_path = token_file

    # Token

    def retrieve_access_token(self):
        """ Read the token from a given file named 'token' """

        token_file = open(self.token_path, 'r')
        return token_file.read().strip()

    # communities

    def list_communities(self, token=None):
        """
        List all the communities, without any filtering.

        :param token: Optional: B2SHARE API ACCESS token
        :return: list of communities in json
        """

        url = urljoin(self.B2SHARE_URL, 'api/communities')

        if token is None:
            try:
                req = self.__send_get_request(url)
                req.raise_for_status()
            except requests.exceptions.HTTPError as err:
                print(err)

        else:
            payload = {'access_token': token}
            try:
                req = self.__send_get_request(url, params=payload)
                req.raise_for_status()
            except requests.exceptions.HTTPError as err:
                print(err)

        if req.status_code == 200:
            return req.json()

    def retrieve_community_specific_records(self, community_id):
        """
        List all records of a specific community.

        :param community_id: community id. Can be retrieved from the list of
        communities
        :exc:`~ecasb2share.ecasb2shareclient.EcasShare.list_communities`
        :return:  the list of records (in JSON format) or an exception message
        """

        payload = {'q': 'community:' + community_id}
        url = urljoin(self.B2SHARE_URL, '/api/records/')

        try:
            req = self.__send_get_request(url,
                                          params=payload)
            req.raise_for_status()
            number_of_records = req.json()['hits']['total']
            print("Total number of records in this community: {}".format(number_of_records))
            if number_of_records > 0:
                return req.json()
            else:
                print("No records in this community")
        except requests.exceptions.HTTPError as err:
            print(err)

    def get_community_schema(self, community_id):
        """
        Retrieves the JSON schema of records approved by a specific community.

        :param community_id: community id. Can be found from the list of
        communities
                            :exc:`~ecasb2share.ecasb2shareclient.EcasShare.list_communities`
        :return: community schema in json format.
        """

        base = self.B2SHARE_URL
        url = '/api/communities/' + community_id + '/schemas/last'

        try:
            req = self.__send_get_request(urljoin(base, url))
            req.raise_for_status()
        except requests.exceptions.HTTPError as err:
            print(err)

        if req.status_code == 200:
            return req.json()

    # records

    def list_all_records(self, size=None):
        """
        List all the records, without any filtering
        TODO add pagination
        :param size: Optional
        :return: list of records in json format.
        """

        url = urljoin(self.B2SHARE_URL, 'api/records')
        if size:
            payload = {'size': size, 'page': 1}
        else:
            payload = {'size': 10, 'page': 1}

        try:
            req = self.__send_get_request(url, params=payload)
            req.raise_for_status()
            return req.json()
        except requests.exceptions.HTTPError as err:
            print(err)

    def get_specific_record(self, record_id, draft=True):
        """ List the metadata of the record specified by RECORD_ID.

        :param record_id: record id.
        :param draft: True/False to specify which type of record to search for.
               Default: True.
        :return: list of records.

         """

        header = {'Content-Type': 'application-json'}
        token = self.retrieve_access_token().rstrip()
        payload = {'access_token': token}

        if draft:
            url = urljoin(self.B2SHARE_URL, 'api/records/' + record_id +
                          '/draft')
        else:
            url = urljoin(self.B2SHARE_URL, 'api/records/' + record_id)

        try:
            req = self.__send_get_request(url, params=payload, headers=header)
            if req is not None:
                req.raise_for_status()
                return json.loads(req.text)
        except requests.exceptions.HTTPError as err:
            print(err)

    def get_record_pid(self, record_id):
        """
        Get the pid from the record metadata (published).

        :param record_id: record id
        :return: epicPID, prefix/suffix
        """

        record = self.get_specific_record(record_id)

        try:
            "{ePIC_PID}".format(**record['metadata'])
            return record['metadata']['ePIC_PID']
        except KeyError as e:
            msg = " missing {} key".format(e)
            raise exceptions.MetadataKeyMissingException(msg=msg)

    def create_draft_record(self, community_id, title):
        """
        Create a new record with minimal metadata, in the draft state.

        :param community_id:
        :param title: title for the record
        :return: record_id, filebucket_id
        """

        token = self.retrieve_access_token().rstrip()
        header = {"Content-Type": "application/json"}
        payload = {'access_token': token}
        metadata = {"titles": [{"title": title}],
                    "community": community_id,
                    "open_access": True}
        url = urljoin(self.B2SHARE_URL, '/api/records/')
        try:
            req = self.__send_post_request(url,
                                           data=json.dumps(metadata),
                                           params=payload,
                                           headers=header)

            req.raise_for_status()
            record_id = req.json()
            filebucket_id = self.get_filebucketid_from_record(record_id['id'])
            print("Draft record created:\n" + record_id['id'])
            print('filebucketid:\n' + filebucket_id)

        except requests.exceptions.HTTPError as err:
            print(err)

        if req.status_code == 201:
            logging.info("Draft record successfully created!")
            return record_id['id'], filebucket_id

    def create_draft_record_with_pid(self, title=None, original_pid=None,
                                     metadata_json=None):
        """

        Create a draft record and specifying the original pid.
        Adapted from DataCite schema:
           * relatedIdentifierType: Handle
           * relationType: isDerivedFrom

        :param title: title for the record.
        :param original_pid: PID (prefix/suffix) of the input Dataset.
        :return: record_id and filebucket_id
        """

        ECAS_COMMUNITY_ID = 'd2c6e694-0c0a-4884-ad15-ddf498008320'
        token = self.retrieve_access_token().rstrip()
        header = {"Content-Type": "application/json"}
        payload = {"access_token": token}
        url = urljoin(self.B2SHARE_URL, '/api/records/')

        if metadata_json:
            metadata = self.load_metadata_from_json(metadata_json)
            self.validate_metadata(metadata)

            related_identifiers = metadata['related_identifiers']

            for pid in range(len(related_identifiers)):
                self.check_pid_syntax(
                    related_identifiers[pid]['related_identifier'])

            try:
                req = self.__send_post_request(url,
                                               data=json.dumps(metadata),
                                               params=payload,
                                               headers=header)

                req.raise_for_status()
            except requests.exceptions.HTTPError as err:
                print(err)

        else:
            metadata = {"titles": [{"title": title}],
                        "community": ECAS_COMMUNITY_ID,
                        "related_identifiers": [
                            {
                                "related_identifier": original_pid,
                                "related_identifier_type": "Handle",
                                "relation_type": "IsDerivedFrom"
                            }
            ],
                "open_access": True
            }
            try:
                req = self.__send_post_request(url,
                                               data=json.dumps(metadata),
                                               params=payload,
                                               headers=header)

                req.raise_for_status()
            except requests.exceptions.HTTPError as err:
                print(err)

        record_id = req.json()
        filebucket_id = self.get_filebucketid_from_record(record_id['id'])
        print("Draft record created:\n" + record_id['id'])
        print('filebucketid:\n' + filebucket_id)

        if req.status_code == 201:
            logging.info("Draft record successfully created!")
            return record_id['id'], filebucket_id

    def submit_draft_for_publication(self, record_id):
        """

        :param record_id: record id
        :return: request status (HTTP response)
        """

        header = {'Content-Type': 'application/json-patch+json'}
        commit = '[{"op": "add", "path": "/publication_state", "value": "submitted"}]'
        token = self.retrieve_access_token().rstrip()
        url = self.B2SHARE_URL + record_id + "/draft"
        payload = {"access_token": token}

        try:
            req = requests.patch(
                url, data=commit, params=payload, headers=header)
            req.raise_for_status()
        except requests.exceptions.HTTPError as err:
            print(err)

        return req.status_code

    def delete_draft_record(self, record_id):
        """

        :param record_id: record id
        :return: request status
        """

        url = urljoin(self.B2SHARE_URL, '/api/records/' + record_id + '/draft')
        token = self.retrieve_access_token().rstrip()
        payload = {'access_token': token}
        header = {"Content-Type": "application/json"}

        req = requests.delete(url, params=payload, headers=header)
        logging.info(req.status_code)
        return req.status_code

    def delete_published_record(self, record_id):
        """
        Notes: only a site administrator can delete a published record.

        :param record_id: the id of the draft record to delete
        :return: request status
        """

        url = urljoin(self.B2SHARE_URL, '/api/records/' + record_id)
        token = self.retrieve_access_token().rstrip()
        payload = {'access_token': token}
        header = {"Content-Type": "application/json"}

        req = requests.delete(url, params=payload, headers=header)

        return req.status_code

    def search_records(self):
        """
        List all the records, without any filtering

        """

        token = self.retrieve_access_token().rstrip()
        header = {"Content-Type": "application/json"}
        payload = {'access_token': token}

        url = urljoin(self.B2SHARE_URL, '/api/records')
        req = self.__send_get_request(url,
                                      params=payload, headers=header)
        return req.json()

    def get_filebucketid_from_record(self, record_id):
        """
        TODO add exception when record not found

        :param record_id:  identifier for a specific record, which
         can be in draft or published state
        :return: filebucket id.
        """

        record = self.get_specific_record(record_id)

        if record is not None:
            filebucket = record["links"]["files"].split('/')[-1]
            return filebucket

    def search_drafts(self):
        """
        Search for all drafts (unpublished records) that are accessible
        by the requestor.
        Usually this means own records only.

        :return: the list of matching drafts (in JSON format).
        """

        token = self.retrieve_access_token().rstrip()
        header = {"Content-Type": "application/json"}
        payload = {'draft': 1, 'access_token': token}
        url = urljoin(self.B2SHARE_URL, '/api/records/?drafts')

        req = self.__send_get_request(url,
                                      params=payload,
                                      headers=header)

        result = json.loads(req.text)
        print(result["hits"]["total"])
        return result

    def search_specific_record(self, search_value):

        payload = {'q': search_value}
        url = urljoin(self.B2SHARE_URL, '/api/records')

        req = self.__send_get_request(url,
                                      params=payload)
        return req.json()

    # files

    def add_file_to_draft_record(self, file_path, filebucket_id):
        """

        :param file_path: path to the file to be uploaded.
        :param filebucket_id: identifier for a set of files.
               Each record has its own file set, usually found
               in the links -> files section


        :return: request status
        """

        header = {'Accept': 'application/json', 'Content-Type': 'octet-stream'}
        token = self.retrieve_access_token().rstrip()
        payload = {'access_token': token}
        upload_file = {"file": open(file_path, 'rb')}
        file_name = os.path.basename(file_path)
        url = urljoin(self.B2SHARE_URL, '/api/files/' + filebucket_id)

        req = self.__send_put_request(url + '/' + file_name,
                                      files=upload_file,
                                      params=payload,
                                      headers=header)

        return req.json()

    def list_files_in_bucket(self, filebucket_id):
        """
        List the files uploaded into a record object.

        :param filebucket_id: identifier for a set of files.
               Each record has its own file set, usually found in
               the links -> files section

        :return: information about all the files in the record object
        """

        token = self.retrieve_access_token()
        payload = {'access_token': token}

        if filebucket_id:
            url = urljoin(self.B2SHARE_URL, '/api/files/' + filebucket_id)
            try:
                req = self.__send_get_request(url, params=payload)
                return req.json()
            except HTTPError as err:
                msg = '{empty request}'.format(err)
                raise exceptions.MetadataKeyMissingException(msg=msg)
        else:
            print("Filebucket ID is None!")
    # requests

    @staticmethod
    def __send_get_request(url, params=None, headers=None):

        s = Session()
        REQUEST_METHOD = 'GET'

        # Build the request
        _request = Request(REQUEST_METHOD, url, params=params, headers=headers)
        prepared_request = _request.prepare()

        try:
            response = s.send(prepared_request)

            # If the response was successful, no Exception will be raised
            response.raise_for_status()
        except HTTPError as http_err:
            print('HTTP error occurred:' + str(http_err))
        except Exception as err:
            print('Other error occurred: ' + err)
        else:
            return response

    @staticmethod
    def __send_put_request(url, files, params, headers):

        s = Session()

        REQUEST_METHOD = 'PUT'

        # Build the request
        _request = Request(REQUEST_METHOD, url, files=files,
                           params=params, headers=headers)
        prepared_request = _request.prepare()

        try:
            response = s.send(prepared_request)

        # If the response was successful, no Exception will be raised
            response.raise_for_status()
        except HTTPError as http_err:
            print('HTTP error occurred:' + http_err)
        except Exception as err:
            print('Other error occurred: ' + err)
        else:
            logging.info('Success!')
            return response

    @staticmethod
    def __send_post_request(url, data, params, headers):

        s = Session()

        REQUEST_MEHOD = 'POST'

        # Build he request
        _request = Request(REQUEST_MEHOD, url, data=data,
                           params=params, headers=headers)
        prepared_request = _request.prepare()

        try:
            response = s.send(prepared_request)
        # If the response was successful, no Exception will be raised
            response.raise_for_status()
        except HTTPError as http_err:
            print('HTTP error occurred:' + http_err)
        except Exception as err:
            print('Other error occurred: ' + err)
        else:
            logging.info('Record created!')
            return response

    @staticmethod
    def __response_status(response):

        if response:
            print("Success!")
        else:
            print('An error has occured.')

    # metadata

    @staticmethod
    def load_metadata_from_json(metadata_json=None):
        if metadata_json:
            with open(metadata_json, 'r') as metadata_json:
                return json.load(metadata_json)

    def validate_metadata(self, metadata=None, metadata_file=None):
        """
        Check if mandatory metadata are passed.

        :param metadata: Optional: metadata as dict (key, value)
        :param metadata_file: Optional. metadata in a json file.
        :return: metadata in json format.
        """

        if metadata_file:
            metadata = self.load_metadata_from_json(metadata_json=metadata_file)

        try:
            "{titles} {related_identifiers} {community} {open_access}".format(**metadata)
        except KeyError as e:
            msg = "missing {} key".format(e)
            raise exceptions.MetadataException(msg=msg)

        return metadata

    @staticmethod
    def check_pid_syntax(pid):
        """
        Checks if the syntax of the Handle is correct.
        Handles should be in this form:
             prefix/suffix

        :param pid: Value of the handle to be checked
        :raise: :exc:`~ecasb2share.exceptions.PIDSyntaxException`
        :return: True, otherwise, exception raised.
        """

        correct_syntax = 'prefix/suffix'

        try:
            arr = pid.split('/')
        except AttributeError:
            raise exceptions.PidSyntaxException(
                msg='The provided handle is None',
                correct_syntax=correct_syntax)

        if len(arr) < 2:
            msg = 'No slash'
            raise exceptions.PidSyntaxException(
                msg=msg, pid=pid, correct_syntax=correct_syntax)

        if len(arr[0]) == 0:
            msg = 'Empty prefix'
            raise exceptions.PidSyntaxException(
                msg=msg, pid=pid, correct_syntax=correct_syntax)

        if len(arr[1]) == 0:
            msg = 'Empty suffix'
            raise exceptions.PidSyntaxException(
                msg=msg, pid=pid, correct_syntax=correct_syntax)

        return True
